sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_client_manager_base.py

```diff
@@ -25,24 +26,106 @@
 MINIMAL_CLIENT_REFRESH_INTERVAL = 30  # 30 seconds
 
+JSON = Mapping[str, Any]
+
 
 @dataclass
 class _ConfigurationClientWrapperBase:
     endpoint: str
 
+    @staticmethod
+    def _generate_allocation_id(feature_flag_value: Dict[str, JSON]) -> Optional[str]:
+        """
+        Generates an allocation ID for the specified feature.
+        seed=123abc\ndefault_when_enabled=Control\npercentiles=0,Control,20;20,Test,100\nvariants=Control,standard;Test,special # pylint:disable=line-too-long
+
+        :param  Dict[str, JSON] feature_flag_value: The feature to generate an allocation ID for.
+        :rtype: str
+        :return: The allocation ID.
+        """
+
+        allocation_id = ""
+        allocated_variants = []
+
+        allocation: Optional[JSON] = feature_flag_value.get("allocation")
+
+        if allocation:
+            # Seed
+            allocation_id = f"seed={allocation.get('seed', '')}"
+
+            # DefaultWhenEnabled
+            if "default_when_enabled" in allocation:
+                allocated_variants.append(allocation.get("default_when_enabled"))
+
+            allocation_id += f"\ndefault_when_enabled={allocation.get('default_when_enabled', '')}"
+
+            # Percentile
+            allocation_id += "\npercentiles="
+
+            percentile = allocation.get("percentile")
+
+            if percentile:
+                percentile_allocations = sorted(
+                    (x for x in percentile if x.get("from") != x.get("to")),
+                    key=lambda x: x.get("from"),
+                )
+
+                for percentile_allocation in percentile_allocations:
+                    if "variant" in percentile_allocation:
+                        allocated_variants.append(percentile_allocation.get("variant"))
+
+                allocation_id += ";".join(
+                    f"{pa.get('from')},"
+                    f"{base64.b64encode(pa.get('variant').encode('utf-8')).decode('utf-8')},"
+                    f"{pa.get('to')}"
+                    for pa in percentile_allocations
+                )
+        else:
+            allocation_id = "seed=\ndefault_when_enabled=\npercentiles="
+
+        # Variants
+        allocation_id += "\nvariants="
+
+        variants_value = feature_flag_value.get("variants")
+        if variants_value and (isinstance(variants_value, list) or all(isinstance(v, dict) for v in variants_value)):
+            if allocated_variants:
+                if isinstance(variants_value, list) and all(isinstance(v, dict) for v in variants_value):
+                    sorted_variants: List[Dict[str, Any]] = sorted(
+                        (v for v in variants_value if v.get("name") in allocated_variants),
+                        key=lambda v: v.get("name"),
+                    )
+
+                    for v in sorted_variants:
+                        allocation_id += f"{base64.b64encode(v.get('name', '').encode('utf-8')).decode('utf-8')},"
```

Comment:
encode defaults to `utf-8`. Prefer to leave it out here. If you prefer being explicit, we should add it to line 115 as well.